### PR TITLE
OCI: Add pagination on `/v2/<name>/tags/list`

### LINKED ISSUE #1936: HTTP API Tags Paginated not work
Hi, I try to get tags with `/v2/<name>/tags/list?n=<integer>`, I have a images that have many tags, when I try to fetch tags with `n`, It return me all the tags, and it also do not give me a `Link` Headers.

My Registry version is `2.5`

### LINKED ISSUE #2452: API different responses
Hi! 
We have a problem with registry (ver. 2.6.1)

Registry API returns manifests and tags (for endpoint - /tags/list). Sometimes we get different responses. The way to reproduce is to repeatedly make requests to API endpoint, and you will get different answers. But during this tests, we didn't change registry state (push and delete action were not performed).

### LINKED ISSUE #3232: OCI Conformance: GET number of tags should be limitable by `n` query parameter
epic: https://github.com/docker/distribution/issues/3203

### Content Discovery > Listing tags

**GET number of tags should be limitable by `n` query parameter**

To resolve this issue, please see the open pull request [3143](https://github.com/docker/distribution/pull/3143). The returned list of tags must be in lexical order and thus deliver consistend results when the `n` parameter is supplied.

```
2020/07/10 06:16:25.228733 DEBUG
==============================================================================
~~~ REQUEST ~~~
GET  /v2/oci-conformance/distribution-test/tags/list?n=2  HTTP/1.1
HOST   : distribution:5000
HEADERS:
	User-Agent: distribution-spec-conformance-tests
BODY   :
***** NO CONTENT *****
-------------------------------------------------------------2020/07/10 06:16:25.230803 DEBUG
==============================================================================
~~~ REQUEST ~~~
GET  /v2/oci-conformance/distribution-test/tags/list?n=2  HTTP/1.1
HOST   : distribution:5000
HEADERS:
	User-Agent: distribution-spec-conformance-tests
BODY   :
***** NO CONTENT *****
------------------------------------------------------------------------------
~~~ RESPONSE ~~~
STATUS       : 200 OK
RECEIVED AT  : 2020-07-10T06:16:25.230528783Z
TIME DURATION: 1.260204ms
HEADERS      :
	Content-Length: 86
	Content-Type: application/json
	Date: Fri, 10 Jul 2020 06:16:25 GMT
	Docker-Distribution-Api-Version: registry/2.0
	X-Content-Type-Options: nosniff
BODY         :
{
   "name": "oci-conformance/distribution-test",
   "tags": [
      "test1",
      "test0",
      "test2",
      "test3"
   ]
}

==============================================================================

2020/07/10 06:16:25.232259 DEBUG
==============================================================================
~~~ REQUEST ~~~
GET  /v2/oci-conformance/distribution-test/tags/list?last=test0&n=2  HTTP/1.1
HOST   : distribution:5000
HEADERS:
	User-Agent: distribution-spec-conformance-tests
BODY   :
***** NO CONTENT *****
------------------------------------------------------------------------------
~~~ RESPONSE ~~~
STATUS       : 200 OK
RECEIVED AT  : 2020-07-10T06:16:25.231990686Z
TIME DURATION: 1.130003ms
HEADERS      :
	Content-Length: 86
	Content-Type: application/json
	Date: Fri, 10 Jul 2020 06:16:25 GMT
	Docker-Distribution-Api-Version: registry/2.0
	X-Content-Type-Options: nosniff
BODY         :
{
   "name": "oci-conformance/distribution-test",
   "tags": [
      "test1",
      "test0",
      "test2",
      "test3"
   ]
}

==============================================================================

Expected
    <int>: 4
to be <=
    <int>: 2-----------------
~~~ RESPONSE ~~~
STATUS       : 200 OK
RECEIVED AT  : 2020-07-10T06:16:25.228472177Z
TIME DURATION: 1.173503ms
HEADERS      :
	Content-Length: 86
	Content-Type: application/json
	Date: Fri, 10 Jul 2020 06:16:25 GMT
	Docker-Distribution-Api-Version: registry/2.0
	X-Content-Type-Options: nosniff
BODY         :
{
   "name": "oci-conformance/distribution-test",
   "tags": [
      "test1",
      "test0",
      "test2",
      "test3"
   ]
}

==============================================================================

Expected
    <int>: 4
to equal
    <int>: 2
```

### PR DESCRIPTION
**Pull request by eyJhb**
Adding pagination to the registry, which is required to live up to the OCI Distribution Specification.
This will also fix a bug with tags being returned in random orders.

Now supports pagination if the user requests it, otherwise return all tags.

fixes #3232 OCI Conformance: GET number of tags should be limitable by n query parameter
closes #1936
closes #2452.

### CODE REVIEW

File: registry/handlers/pagination.go [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+	"net/url"
+	"strconv"
+
+	"github.com/docker/distribution/registry/api/errcode"
+	"github.com/docker/distribution/registry/storage/driver"
+)
+
+const maximumReturnedEntries = 100
+
+type paginateStruct struct {
```
Comment (eyJhb) [inline registry/handlers/pagination.go:18]: Ensure about the struct, but the function signature is very long otherwise.

**Comment by eyJhb** [comment]
Simplified the code for adding pagination.
It now supports pagination if the user requests it, otherwise return all tags.

Anyone who wants to look at it is welcome!

**Comment by eyJhb** [comment]
@dmcgowan  should be good to go! Have a look at it when you get the time.
Should be the last blocker before a release could be made (as far as I understood from comments).

File: registry/handlers/tags.go [hunk 2]
```diff
@@ -49,6 +51,39 @@ func (th *tagsHandler) GetTags(w http.ResponseWriter, r *http.Request) {
...
+	// get entries after latest, if any specified
+	if lastEntry := q.Get("last"); lastEntry != "" {
+		lastEntryIndex := sort.SearchStrings(tags, lastEntry)
+		if lastEntryIndex != len(tags) && tags[lastEntryIndex] == lastEntry {
+			tags = tags[lastEntryIndex+1:]
+		}
+	}
+
+	// if no error, means that the user requested `n` entries
+	if maxEntries, err := strconv.Atoi(q.Get("n")); err == nil {
```
Comment (dmcgowan) [inline registry/handlers/tags.go:65]: I think it makes more sense here to return 400 if n is a garbage value rather than just ignore it. Just go into this block if `n` is set.

Comment (eyJhb) [reply registry/handlers/tags.go:65]: Sure, I can do that! 
Keep in mind that atm. catalog has the same behavior, where it will not throw any error on a invalid `n` value (https://github.com/docker/distribution/blob/master/registry/handlers/catalog.go#L41-L44).

Should I make a PR to get them to match up?

Comment (dmcgowan) [reply registry/handlers/tags.go:65]: I think that is the right behavior, but considering that API is not used by end clients, it's not super important.

We can possibly discuss the error handling of these in the distribution spec as well if it is too ambiguous. Handling bad input (n is not a number, last does not exist) could be clearly defined.

Comment (eyJhb) [reply registry/handlers/tags.go:65]: Lets get this merged first then I can get the other branch up-to-date and use the distribution spec in that as well.
Somewhat easy to do for `n`, but when `last` is wrong it is a different case (either the tag does not exists or the repository doesn't).

I currently have this wording for the spec, but feel unsure about the wording.

```go
	// ErrorPaginationNumberInvalid is returned when the requested number
	// of results n is not a integer.
	ErrorPaginationNumberInvalid = errcode.Register(errGroup, errcode.ErrorDescriptor{
		Value:   "PAGINATION_NUMBER_INVALID",
		Message: "invalid number of results requested",
		Description: `This error may be returned when the requested
		"n" number of results is not an integer type`,
		HTTPStatusCode: http.StatusBadRequest,
	})
```

Would love some input, and maybe a suggestion for the `last` param. Else I will try to word it myself tomorrow :)

Comment (eyJhb) [reply registry/handlers/tags.go]: @dmcgowan fixed in latest commits, please check to ensure it is correct :)

Comment (eyJhb) [reply registry/handlers/tags.go]: Hoping a ping is OK,new Github notifications isn't nice @dmcgowan :|

Comment (dmcgowan) [reply registry/handlers/tags.go]: I saw this, but wanted to cross reference these changes with the spec first before merging. Just went through now, the section is https://github.com/opencontainers/distribution-spec/blob/master/spec.md#tags-paginated

File: registry/handlers/tags.go [hunk 3]
```diff
@@ -49,6 +51,48 @@ func (th *tagsHandler) GetTags(w http.ResponseWriter, r *http.Request) {
...
 
+	// do pagination if requested
+	q := r.URL.Query()
+	// get entries after latest, if any specified
+	if lastEntry := q.Get("last"); lastEntry != "" {
+		lastEntryIndex := sort.SearchStrings(tags, lastEntry)
+		if lastEntryIndex != len(tags) && tags[lastEntryIndex] == lastEntry {
+			tags = tags[lastEntryIndex+1:]
+		} else {
+			th.Errors = append(th.Errors, v2.ErrorCodePaginationLastUnknown.WithDetail(map[string]string{"last": lastEntry}))
```
Comment (dmcgowan) [inline registry/handlers/tags.go:62]: This error is the one that I wanted to double check. The spec says for `last` that  `Result set will include values lexically after last`. In this case, the search should be done, but it can still continue even if last doesn't exist but entries exist lexically after it.

@jdolitsky do you mind offering an opinion here?

Comment (eyJhb) [reply registry/handlers/tags.go:62]: That makes perfect sense tbh. most have missed that it specified that. I sadly read it as after the index of the `last` string and values should be sorted lexically.
Seems like that should be the behavior :)

Just say the word and I will update it in this (and update the catalog PR after this is merged to do the same).

Comment (pmengelbert) [reply registry/handlers/tags.go:62]: @dmcgowan I'm here on behalf of @jdolitsky
You're correct, the spec says that  the `/v2/<name>/tags/list` endpoint should return a lexically-ordered list when the `n` query parameter is present, regardless of whether `last` is specified.

Usually the value of the `last` key would be acquired by by a `Link` header from an earlier response. Failing that, the registry should probably behave intelligently when the actual `last` value doesn't exist.

The current language in the spec is at the SHOULD level, but this will probably become a MUST by the time distribution-spec 1.0 is released.

Comment (eyJhb) [reply registry/handlers/tags.go:62]: What I gather from this is the following. `last` does not have to match a value in the tags, but should specify where in the tags we want to get our tags from legically (so this should be changed in this PR).

I am wondering if `last` should be ignored, if no `n` is specified? Below is from the doc.

```
The client can then issue the request with the above value from the Link header, receiving the values v3 and v4. Note that n may change on the second to last response or be fully omitted, depending on the server implementation.
```

Keep in mind, this will also change the `catalog` listing to work the same way as `tags`.

Comment (jdolitsky) [reply registry/handlers/tags.go:62]: This seems related to this discussion: https://github.com/opencontainers/distribution-spec/issues/132

Comment (eyJhb) [reply registry/handlers/tags.go:62]: Okay, so what I think should happen is the following.

- `n=0` will return a empty list, and will not include a `Link`, as there are no more entries (depends on how you look at it. But if I want 0 and I want more, I will still get 0).
- `last` will be used to find the entries lexically after it, and the index does not need to exists.
    - this could bring some weird bugs for some people regarding encoding. but it does make sense to be able to search like this.
- `last` can stand on its own without `n`, as to use the default `n` value from the server (which it may impose).

How does this sound @jdolitsky and @dmcgowan ?

Comment (dmcgowan) [reply registry/handlers/tags.go:62]: Makes sense to me, but we should probably clarify in the spec first. I'm ok making that update in this project with expectation that we may need to make slight changes if we come to a different agreement for the spec

Comment (eyJhb) [reply registry/handlers/tags.go:62]: @dmcgowan  so this is good to go, if I make the changes and then we can update the spec in accordance?

Comment (dmcgowan) [reply registry/handlers/tags.go:62]: yeah, let's go ahead and make those changes and get this merged in. We can tweak the conformance tests and spec after, if we need to come back and sync here we can before the release

Comment (eyJhb) [reply registry/handlers/tags.go]: @dmcgowan  thougt I had pinged you here, I guess not. It should be good to go! Travis is giving some weird errors, that doesn't seem to be a problem with the code. Should be ready to merge!

**Comment by eyJhb** [comment]
Anyone that can merge this? Seems to have been forgotten :)

**Comment by thaJeztah** [comment]
Looks like the build was failing; I restarted CI to see if it was a one-off failure


@manishtomar @dmcgowan PTAL

**Comment by eyJhb** [comment]
@thaJeztah the errors does not seem related to my work, or am I reading it wrong?

```
+ check

/bin/sh: 1: golangci-lint: not found

make: *** [check] Error 127

The command "make check" exited with 2.
```

**Comment by jdolitsky** [comment]
@eyJhb @thaJeztah - the CI failure is unrelated to the change, I have the same Travis error in an unrelated PR: https://github.com/docker/distribution/pull/3161

**Comment by thaJeztah** [comment]
Ah, thanks I see @jdolitsky opened https://github.com/docker/distribution/pull/3187 to fix CI

**Comment by eyJhb** [comment]
So, waiting for #3187 so I can cherry pick it. Assume it won't take long

**Comment by thaJeztah** [comment]
@eyJhb https://github.com/docker/distribution/pull/3187 was merged; can you rebase your PR to get that fix in? Let me know if you need a hand / instructions for the rebase

**Comment by eyJhb** [comment]
> @eyJhb #3187 was merged; can you rebase your PR to get that fix in? Let me know if you need a hand / instructions for the rebase

Rebased and passed! Ready to be merged

File: registry/handlers/tags.go [hunk 4]
```diff
@@ -49,6 +51,43 @@ func (th *tagsHandler) GetTags(w http.ResponseWriter, r *http.Request) {
 		return
 	}
 
+	// do pagination if requested
+	q := r.URL.Query()
+	// get entries after latest, if any specified
+	if lastEntry := q.Get("last"); lastEntry != "" {
+		lastEntryIndex := sort.SearchStrings(tags, lastEntry)
+		tags = tags[lastEntryIndex+1:]
```
Comment (dmcgowan) [inline registry/handlers/tags.go:59]: According to the documentation for [`sort.SearchStrings`](https://golang.org/src/sort/search.go?s=3680:3724#L91), `x is not present (it could be len(a))`. This could lead to a panic here.  This should probably be covered in the unit tests and boundary tested.

Comment (eyJhb) [reply registry/handlers/tags.go:59]: I just seem to remember having fixed that error before.
To do unittests, that would require I know a lot more regarding how unit tests are done in the handlers.

I propose making this change (fixing this), so it will not panic and then we can either merge and someone can make a PR for unit tests, or there can be made a PR based on this.

But considering this has been in the making since 2016, I am not having high hopes for anyone picking up the unittests. And not sure when I will get the time for it, sadly

Comment (jdolitsky) [reply registry/handlers/tags.go:59]: @eyJhb maybe at least to ensure no panic we can add something like
```go
if lastEntry := q.Get("last"); lastEntry != "" {
	lastEntryIndex := sort.SearchStrings(tags, lastEntry) + 1
	if lastEntryIndex < len(tags) {
		tags = tags[lastEntryIndex:]
	}
}
```
?

Comment (eyJhb) [reply registry/handlers/tags.go]: @jdolitsky ahh damn it, that is a lot prettier than what I just pushed... Lets totally go with that! Let me just rebase again :)

Comment (eyJhb) [reply registry/handlers/tags.go]: EDIT: we can't do that actually, as we need to set it to a empty list if it has no matches. That is at least what I would think if a user specifies something that is not in the tags (return nothing). But we can do this 

What about the newest change?

```go
	if lastEntry := q.Get("last"); lastEntry != "" {
		lastEntryIndex := sort.SearchStrings(tags, lastEntry)

		// as`sort.SearchStrings` can return len(tags), if the
		// specified `lastEntry` is not found, we need to
		// ensure it does not panic when slicing.
		if lastEntryIndex < len(tags) {
			tags = []string{}
		} else {
			tags = tags[lastEntryIndex+1:]
		}
	}
```

Comment (jdolitsky) [reply registry/handlers/tags.go]: Yes, noticed that too. I agree. This lgtm

Comment (eyJhb) [reply registry/handlers/tags.go]: I did however like yours a lot, but we end up having somewhat the same :) Perfect! Lets see if the others agree.
@dmcgowan will you resolve if you think this solution is acceptable?

Comment (eyJhb) [reply registry/handlers/tags.go]: Ping @dmcgowan :)

**Comment by manishtomar** [comment]
Code LGTM. However, I'll run tests on AWS and get back.

File: registry/handlers/tags.go (hunk 4 above)
Comment (eyJhb) [reply registry/handlers/tags.go]: Ping again, hope it is okay @dmcgowan :)

@thaJeztah anything we can do to get this merged?

**Comment by waynr** [comment]
Anything I can do to help get this PR merged? From the comments it looks like everything has been worked out and it is just waiting on a tardy reviewer to give thumbs up.

**Comment by eyJhb** [comment]
> Anything I can do to help get this PR merged? From the comments it looks like everything has been worked out and it is just waiting on a tardy reviewer to give thumbs up.

I think we are just waiting for someone to merge this. At least as far as I can tell.

**Comment by eyJhb** [comment]
Bumping this :)

**Comment by thaJeztah** [comment]
@dmcgowan @manishtomar PTAL 🙏

**Comment by eyJhb** [comment]
Massive ping to everyone, hoping this can get merged soon :)

@dmcgowan  @manishtomar  @thaJeztah

**Comment by eyJhb** [comment]
Ping @dmcgowan @manishtomar :)

**Comment by thaJeztah** [comment]
@arkodg @olegburov ptal

**Comment by thaJeztah** [comment]
@arkodg @olegburov ptal

**Comment by eyJhb** [comment]
Ping @arkodg and @olegburov based on @thaJeztah pings :)

**Comment by deleteriousEffect** [comment]
Since we've introduced a new `ErrorPaginationNumberInvalid` error, I'd prefer that we either add documentation for that either in this MR, or open up a follow-up issue before this is merged.

**Comment by milosgajdos** [comment]
@eyJhb please rebase this PR

**Comment by eyJhb** [comment]
> @eyJhb please rebase this PR

Rebased and should be ready :)

**Review by milosgajdos — CHANGES_REQUESTED** [review]
The code looks good. Thanks @eyJhb for putting this together. There is one change I'd like to make as per @deleteriousEffect comment. Can you please add the newly introduced error code into `docs/spec/api.md`. Then we are good to go!

**Comment by milosgajdos** [comment]
FYI, future reviewers, the spec has moved the tag "pagination" section into content discovery https://github.com/opencontainers/distribution-spec/blob/main/spec.md#content-discovery

**Comment by eyJhb** [comment]
> The code looks good. Thanks @eyJhb for putting this together. There is one change I'd like to make as per @deleteriousEffect comment. Can you please add the newly introduced error code into `docs/spec/api.md`. Then we are good to go!

Should be good to go now I suppose? :)

**Review by milosgajdos — APPROVED** [review]

**Comment by wy65701436** [comment]
can we add UT for this? To test different situation, both positive and negative.

- httptest.NewRequest(http.MethodGet, "/v2/library/hello-world/tags/list", nil)
- httptest.NewRequest(http.MethodGet, "/v2/hello-world/tags/list?n=1", nil)
- httptest.NewRequest(http.MethodGet, "/v2/hello-world/tags/list?n=3", nil)
- httptest.NewRequest(http.MethodGet, "/v2/hello-world/tags/list?last=v1&n=1", nil)

File: registry/handlers/tags.go [hunk 5]
```diff
@@ -49,6 +51,51 @@ func (th *tagsHandler) GetTags(w http.ResponseWriter, r *http.Request) {
...
+	// get entries after latest, if any specified
+	if lastEntry := q.Get("last"); lastEntry != "" {
+		lastEntryIndex := sort.SearchStrings(tags, lastEntry)
+
+		// as`sort.SearchStrings` can return len(tags), if the
+		// specified `lastEntry` is not found, we need to
+		// ensure it does not panic when slicing.
+		if lastEntryIndex == len(tags) {
+			tags = []string{}
+		} else {
```
Comment (wy65701436) [inline registry/handlers/tags.go:65]: if the last doesn't exist, should not be a bad request?

Comment (eyJhb) [reply registry/handlers/tags.go:65]: We can do partial searches, using the lastEntryIndex, so it doesn't need to be a exact match, and therefore is not really a error as far as I understand :)

**Comment by joaodrp** [comment]
> can we add UT for this? To test different situation, both positive and negative.
> 
> * httptest.NewRequest(http.MethodGet, "/v2/library/hello-world/tags/list", nil)
> * httptest.NewRequest(http.MethodGet, "/v2/hello-world/tags/list?n=1", nil)
> * httptest.NewRequest(http.MethodGet, "/v2/hello-world/tags/list?n=3", nil)
> * httptest.NewRequest(http.MethodGet, "/v2/hello-world/tags/list?last=v1&n=1", nil)

@eyJhb, could you please add some tests to `registry/handlers/api_test.go`?

**Comment by eyJhb** [comment]
> @eyJhb, could you please add some tests to registry/handlers/api_test.go?
I am on the last weeks of my semester project, I will try to see once I have time. But if anyone wants to throw some in that have time, feel free :)

**Comment by joaodrp** [comment]
> @eyJhb, could you please add some tests to registry/handlers/api_test.go?
> I am on the last weeks of my semester project, I will try to see once I have time. But if anyone wants to throw some in that have time, feel free :)

Alright, I'll push a separate PR with test coverage. Just tested this locally with no problems.

**Comment by rbbl-dev** [comment]
i just pulled a fresh image of this registry and i'm questioning my sanity.

for me the tags are not sorted and `n` and `last` are ignored but this MR claims to fix those things.
grafik

**Comment by sudo-bmitch** [comment]
@rbbl-dev this is in the 3.0.0 milestone. What version of the registry are you running?

**Comment by rbbl-dev** [comment]
2.8.2 which is the newest on https://hub.docker.com/_/registry/tags @sudo-bmitch
